### ISSUE TITLE
FP16 训练 ResNet50，loss 为负数

### ISSUE THREAD

**Issue by mzchtx** [PaddlePaddle/models#2909]
# 环境

+ system: centos6u3
+ python: 2.7.14
+ paddle: 1.5.1.post97
+ nccl: 2.2.13

# 问题

使用 FP16 训练 ResNet50:

如果在 fc layer 里加上 softmax 激活，然后使用 `cross_entropy` 损失，会出现几个 batch 后 loss 为负数的情况。如下：

```python
out = ResNet50().net(input=image, class_dim=class_dim)
cost = fluid.layers.cross_entropy(input=out, label=label)
avg_cost = fluid.layers.mean(x=cost)
```

![image](https://user-images.githubusercontent.com/37564754/61769557-4071df80-ae1d-11e9-989e-28d5023db4be.png)

如果在 fc layer 不加 softmax 激活，然后使用 `softmax_with_cross_entropy` 损失，则结果正常。如下：

```python
out = ResNet50().net(input=image, class_dim=class_dim)
cost = fluid.layers.softmax_with_cross_entropy(input=out, label=label)
avg_cost = fluid.layers.mean(x=cost)
```

![image](https://user-images.githubusercontent.com/37564754/61769622-757e3200-ae1d-11e9-9a5c-718ebd007287.png)

上述两种方式使用 fp32 进行训练都没问题。

**Comment by shippingwang** [comment]
FP16问题将在未来两周内修复

**Comment by gavin1332** [comment]
softmax和ce拆分以后，ce的梯度需要计算label概率p的倒数，但p较小的时候会导致出现数值计算问题。因为float16的表示范围较float32小，更容易出现underflow的情况，也就是p更容易为0，导致后续计算结果异常

**Comment by gavin1332** [comment]
已和提问者沟通确认次问题。先将此issue关闭
